main.c

```diff
@@ -4619,39 +4612,51 @@ ovl_rename_direct (fuse_req_t req, fuse_ino_t parent, const char *name,
       unlinkat (destfd, newname, 0);
     }
 
-  /* Try to create the whiteout atomically, if it fails do the
-     rename+mknod separately.  */
+  bool src_needs_whiteout = (node->last_layer != get_upper_layer (lo));
+
+  if (src_needs_whiteout)
+    {
+      /* Trying to atomically both rename and create the whiteout.
+         If destination is a whiteout, we can EXCHANGE source and destination and reuse the old whiteout.
+         If not, we can try to atomically create one with the WHITEOUT flag.  */
+      if (destnode_is_whiteout)
+        ret = direct_renameat2 (srcfd, name, destfd, newname, flags|RENAME_EXCHANGE);
+      else
+        {
   if (! can_mknod)
     {
       ret = -1;
       errno = EPERM;
     }
   else
-    {
-      ret = direct_renameat2 (srcfd, name, destfd,
-                              newname, flags|RENAME_WHITEOUT);
+            ret = direct_renameat2 (srcfd, name, destfd, newname, flags|RENAME_WHITEOUT);
```

Comment:
indentation seems off

Reply:
That's VS Code or Git doing something weird. It was right when I committed it. 🙄